### DISCUSSION TITLE
Inverter remote control

### DISCUSSION THREAD

**Discussion by Satal** [hultenvp/solis-sensor#246] [Show and tell]
Hey all,

I'm not sure if this is the right place, but I thought I would share it as I imagine it's something that a lot of you will be interested in.

While submitting a ticket to Solis, I noticed that there was a category to request remote control of your inverter. This allows you to control the inverter from the website or your phone. From a brief look it looks like you can control anything that you could directly from the screen on the inverter.

For me, the main use case is being able to set the charge from the grid without having to climb into my loft. I'm guessing that this is exposed via some API though I doubt it's going to be the same API we're using for reading the information.

To get the Remote Control Access, go to the [Solis Service Center](https://solis-service.solisinverters.com/support/tickets/new) and submit a new ticket specifying the "Product Type" to be "PV Inverter" and the "Tickets Type" to "Remote Control Access", filling out the rest of the details as you would normally. I would recommend being clear in your description that you want Remote Control Access in addition to the API access, as initially, the support team tried adding API access a second time, despite me already having it enabled.

You should then get a response in a day or so letting you know that the Remote Control Access has been approved and you need to log out and log back in.

In the email, you'll get the following screenshots of how to access the Remote Control area.

![01-Device_page](https://user-images.githubusercontent.com/3852030/211189787-d30c28b5-2a8b-4400-b636-1232dd0102e6.png)
![02-Inverter_details](https://user-images.githubusercontent.com/3852030/211189790-af96b642-6531-4bf1-b1dd-98e95ced33fa.png)
![03-Control_Inverter](https://user-images.githubusercontent.com/3852030/211189792-4f67ebcb-9166-457a-a9d4-e1c5193e4fe0.png)
![04-Inverter_control](https://user-images.githubusercontent.com/3852030/211189793-29ffc9aa-c785-45a8-9448-f2a85c962fd7.png)

You will also be able to access it through the website, which looks like this
![05-web](https://user-images.githubusercontent.com/3852030/211189798-c7abacea-a163-44f3-a1b2-5f0c01964e79.png)

Whether you go in through the app or the website, they will present you with a disclaimer that you have to agree to regarding the use of this functionality.

Anyway, I thought I would share and hope it is useful 😊

**Comment by hultenvp** [comment]
Interesting read @Satal!

At this moment there's no indication in the API v1.2 that they allow sending instructions to the inverter.
Did you happen to get access to additional or newer API documentation?

**Reply by Satal** [reply]
> Interesting read @Satal!
> 
> At this moment there's no indication in the API v1.2 that they allow sending instructions to the inverter. Did you happen to get access to additional or newer API documentation?

They didn't provide any further documentation than just the screen shots. 

I'll ask them if there is anyway to access this through an API and let you know what they say

**Reply by Satal** [reply]
@hultenvp, after quite a lot of back and forth with the Solis support team, they've said that there isn't an API for Remote Control.

I've had a quick look at what happens in Chrome when you save, and it appears that a call to `https://www.soliscloud.com:15555/inverter/atCommandV3`

```json
{
    "cid":"103",
    "inverterId":"[My Inverter ID]",
    "value":"50,50,00:00,00:00,00:00,00:00,0,0,00:00,00:00,00:00,00:00,0,0,00:00,00:00,00:00,00:00",
    "localTimeZone":0,
    "language":"2"
}
```
The value above appears to be comma separated values for the "Charge and Discharge" settings
- Charging Current1
- Discharging Current1
- Charging Time1 Start
- Charging Time1 End
- Discharging Time1 Start
- Discharging Time1 End
- Charging Current2
- Discharging Current2
- Charging Time2 Start
- Charging Time2 End
- Discharging Time2 Start
- Discharging Time2 End
- Charging Current3
- Discharging Current3
- Charging Time3 Start
- Charging Time3 End
- Discharging Time3 Start
- Discharging Time3 End

The response to the request is:
```json
{
    "success":true,
    "code":"0",
    "msg":"success",
    "data":[
        {
            "msg":"Home Inverter : [My Inverter ID]    Time: 1673868200<br>NO: 176<br>SN: [Inverter Serial Number]<br>Result: 0<br>Inverter1: <br>send: <br>recv: [Receiver number?]<br>",
            "code":0,
            "recv":"[Receiver number?]",
            "command":"AT+TEST=GIN[3 digits]:01 10 [HEX] [HEX] 00 1e 3c 01 f4 01 f4 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 f1 79"
        }
    ]
}
```

That may all mean something to someone with more experience than me.

**Reply by hultenvp** [reply]
Looks a lot like the old Ginlong V2 API, which uses kind of the same request-response pattern for reading data.
I do assume this is a recorded command for one of the settings. Is this "Battery Setting"?

Shouldn't be too difficult to write some code for this, but I do not have the right logger for this myself (same as @alienatedsec )

**Comment by alienatedsec** [comment]
What is your datalogger @Satal ?

```
*Also please note if you are using a DLS-W, DLS-L or RF-LINK datalogger we recommend to upgrade to S3-WIFI-ST, S2-WL-ST or S3-GPRS-ST for better functionality.  

If you get an error showing "Data error please contact webmaster" you absolutely need to upgrade your logger.  
```

Mine is DLS-W running `MW3_15_0501_1.18` and I had the following error as described above
![image](https://user-images.githubusercontent.com/73167064/211303974-cbd23726-2f8d-48fd-9a9b-043f47640663.png)

**Reply by Satal** [reply]
> What is your datalogger @Satal ?
> 
> ```
> *Also please note if you are using a DLS-W, DLS-L or RF-LINK datalogger we recommend to upgrade to S3-WIFI-ST, S2-WL-ST or S3-GPRS-ST for better functionality.  
> 
> If you get an error showing "Data error please contact webmaster" you absolutely need to upgrade your logger.  
> ```
> 
> Mine is DLS-W running `MW3_15_0501_1.18` and I had the following error as described above ![image](https://user-images.githubusercontent.com/73167064/211303974-cbd23726-2f8d-48fd-9a9b-043f47640663.png)

I'm not sure how to check without going into my loft, is there somewhere convenient to see it in the app?

I can see the data logger version as 1012f. Solis just did a firmware update for me as I have a ticket open about getting the wrong data when charging from grid, but I had the remote control working before that update.

**Reply by alienatedsec** [reply]
It will be after you log in to your stick directly (more info will be there) or via portal in datalogger section (just firmware version)

![image](https://user-images.githubusercontent.com/73167064/211305474-a193f533-577b-42a7-97fe-c12be6afb8b3.png)

**Reply by Satal** [reply]
Ah OK, thank you.

It looks like the password for the data logger may not be the default. Looks like I'll be going into the loft later and will let you know.

For the firmware, this is what I get through the app
![Screenshot_20230109-121441~2](https://user-images.githubusercontent.com/3852030/211307775-861addd1-8f54-4377-a757-8c8cf53a36c4.png)

**Reply by alienatedsec** [reply]
Thanks. It's definitely a different version of what DLS-L or DLS-W are, as the firmware format differs greatly.

**Reply by Satal** [reply]
Yeah, just been in the loft and it looks like I have a S3-WIFI-ST

**Reply by magichew** [reply]
I have the same S3-WIFI-ST logger. Has there been any progress with controlling the inverter?

**Reply by alienatedsec** [reply]
Not from this integration, but Solis Cloud allows you remote control the inverter using your data logger. You need to raise a ticket.

**Reply by Satal** [reply]
> have the same S3-WIFI-ST logger. Has there been any progress with controlling the inverte

Admittedly not, I had a baby born on the 18th of Jan, which has taken up most of my time since.

I have used Chrome to capture the request from the SolisCloud website. However, the auth changes regularly so I'm unable to use that regularly and I haven't figured out how to generate the auth code.

**Reply by magichew** [reply]
> Not from this integration, but Solis Cloud allows you remote control the inverter using your data logger. You need to raise a ticket.

Hi yes, I have remote control access enabled on Solis Cloud and API access so all the data is coming in to HA but of course looking to control via HA for the benefits of automating the process.

It's a real shame control isn't supported in the Solis API. It would tidy things up very neatly.

**Reply by alienatedsec** [reply]
@magichew you still have some options on the table - although, you need to invest

https://github.com/alienatedsec/solis-ha-modbus-cloud

**Reply by magichew** [reply]
Thanks for the tip. I've been looking in to it and there seems to be many ways to achieve what I'm looking to do. Guess I need to find the best for me.

Best for me would just be if Solis would allow the API to send commands to the inverter in the same way the inverter can be controlled by the web interface.

**Reply by alienatedsec** [reply]
Both platforms - HA (solis-sensor) integration, and web interface - use the same API - one difference, the reporting API and controlling API have different permissions. Also, the former is published and the latter seems to be isolated with explicit access to Solis solutions only (e.g., remote web interface).

I guess, the API was initially created for data to be managed outside of the Solis Cloud, and for your own benefit.

**Reply by sjoram** [reply]
> @magichew you still have some options on the table - although, you need to invest
> 
> https://github.com/alienatedsec/solis-ha-modbus-cloud

Hey - I've just ordered those parts! I might be looking for some help in a week or so!

**Reply by alienatedsec** [reply]
@sjoram happy to help. Please remember to post any questions in the repo rather than over here.

**Reply by magichew** [reply]
> @sjoram happy to help. Please remember to post any questions in the repo rather than over here.

Does your solution allow for writing back to the inverter to control it via HA to for example automate import from grid to charge batteries at the cheapest 30 min period?

**Reply by alienatedsec** [reply]
@magichew it does allow when combined with the right integration and the ha automation around it. Lots of work, but there are many guides and videos on how to approach it.

**Comment by haydnw** [comment]
Just a note that Solis support today told me that remote control only works on hybrid inverters, not string ones (like ours).

**Reply by alienatedsec** [reply]
@haydnw they forgot to mention that only certain dataloggers are supported when it comes to the remote control. 

https://github.com/alienatedsec/solis-ha-modbus-cloud

**Comment by lynxus** [comment]
Did anyone ever get this working?

When I try to send data to that endpoint, I get the following:

{
    "error": "Unauthorized",
    "message": "illegal resource1300386381676591899",
    "path": "/inverter/atCommandV3",
    "status": 401,
    "timestamp": 1686909288819
}


If anyone ever got a python script to work with this atCommandV3 endpoint, please share.... This is killing me :)


I'm starting to wonder if this endpoint allows "API" authentication.
Copying the CURL and re-running from Chrome works fine, but using the API Authorisation doesn't...

That "Resource" (resource1300386381676591899)   I have no idea what that is... Its not in my headers or body, Its not the inverter ID. etc..


**Wondering if theres a way to "scrape" and pretend the script is a web user and not api....**

**Comment by R-ens** [comment]
Hello,

I noticed that since a few days there is also support for string inverters. At least my model gives an On and Off button to control my Solis inverter remotely! 🎉

So I reached out to the local support team immediately (Netherlands/Belgium) because a few days earlier they enabled the remote control for me. Got a blank page and after some messaging we came to the conclusion that mine was a string converter which isn't supported. So I asked if it was planned in future and he replied with no. He told me later that he also reached out to China (headquarters?)

Apparently there is new software released inbetween those days which makes this feature available. I looked for release notes or whatsoever but couldn't find anything.

**Comment by hultenvp** [comment]
I wonder if we can create a HA service for this, especially now HA services also can return data.

**Reply by ram1t** [reply]
This would be the ultimate if you got this working!

**Comment by bryanajwells** [comment]
Hello All

I too have been interested in remote control of the inverter.  I reached out to Solis UK earlier in the week, by creating a support ticket (https://solis-service.solisinverters.com/en/support/tickets/new), requesting any documentation for the remote control API.
After signing an NDA (which means unfortunately I can not share the document with everyone) they send me a document entitled:

SolisCloud Platform API Document - Device Control V2.0

Sections 1 and 2 of 'my' document appear to be the same as sections 1 and 2 of the document (SolisCloud Platform API Document V2.0) stored in the repository, leaving only about 6 pages of "new" stuff.
This would imply that the general usage is pretty much the same as the the now, so extending the module is probably doable without massive churn.

The document refers out to a s/sheet containing a list of data items it is possible to set/retrieve.  This lists things like charge times/currents etc.  i.e. the same items that can been seen on the app's Inverter Control section.

Maybe if more of us reach out in a similar way, maybe they can be pursuaded to release the document without the NDA?


Meanwhile I am going to reachout to them again requesting permission under the NDA to share the document with the dev-team, for the purposes of updating the plugin and distributing.

**Reply by FrankTaylorLieder** [reply]
This is very interesting. I am currently getting quotes from an installer for a system including a Solis Hybrid inverter. If there is API access to control battery charging I will push ahead with the install and if possible help out any work here to integrate.

I am looking to control the battery charging using a solar prediction for the following day to determine how much of the battery to charge on cheap rate overnight power and how much to leave for solar the following day. Has anyone else done this for other systems?

**Reply by alienatedsec** [reply]
@FrankTaylorLieder there are plenty of ways to manage your battery but the API is not the best to handle it. Please check the following:

https://github.com/fboundy/ha_solis_overview << connectivities explained
https://github.com/fboundy/pv_opt << Battery optimisation
https://github.com/alienatedsec/solis-ha-modbus-cloud << a connectivity method to HA

**Reply by bryanajwells** [reply]
As a software engineer doing it entirely in software intreagues me. Just in the same way as I haven't spent money on a zappi, but have developed with the help of somewhere else on the interwebs cant remeber where though, a Home-assistant integration to charge the car on the excess - after the battery and hot water are fully topped up.   To date though I haven't managed to even login (must dedicate more effort to this project).

I have seen solutions like the bottom link which requires extra hardware to implement, however I am not yet prepared to go to that yet.  However it would give more realtime information.  

While I like the inverter and battery setup, one feels that the 3S-Wifi-ST has been produced too cheaply and takes out of our hands things which might have been useful.  SolisCloud needs the ability to get the data more frequently.  The API shouldn't be behind a NDA.

Then again SolisCloud and all your data goes via China, and what if SolisCloud goes down.
(I own an Anki Vector and until last week he had become dumb because the DDL servers are borked.  Escape Podded him.)

**Reply by alienatedsec** [reply]
> Then again SolisCloud and all your data goes via China, and what if SolisCloud goes down.
> (I own an Anki Vector and until last week he had become dumb because the DDL servers are borked. Escape Podded him.)

@bryanajwells hence, why I made what I made and feel completely independent - I was away in December; however, noticed in January that my Datalogger is down and not reporting to SolisCloud - not bothered at all as all my data was, and still is, and will be in Home Assistant.

**Reply by bryanajwells** [reply]
For those following this thread - I found this today quite by chance.  I was looking for something else.
https://github.com/stevegal/solis_control
Not tried it yet but an interesting development none the less.

**Comment by JanVanC** [comment]
Hi,

I'm also very interested in this. I'm new to HA and I think it would be really useful to set certain parameters of the inverter. Mainly to decide when to charge the battery from the grid and when to allow discharge of the battery.

It looks like the inverter control has recently been rolled out to everyone in soliscloud.com. I wonder if this means that it can now also be changed over the API.

I'm not familiar with plugin development, but I'm happy to help in testing some functionality.

**Comment by ausloki** [comment]
I applied for access to the API but after two weeks I still hadn’t received a response. I responded to the auto generated ticket and then got a response. They wanted me to fill out a non disclosure agreement.

After reading through all the fine print gave it up


Sent from my iPhone

On 26 Jan 2024, at 2:02 am, FrankTaylorLieder ***@***.***> wrote:

﻿

This is very interesting. I am currently getting quotes from an installer for a system including a Solis Hybrid inverter. If there is API access to control battery charging I will push ahead with the install and if possible help out any work here to integrate.

I am looking to control the battery charging using a solar prediction for the following day to determine how much of the battery to charge on cheap rate overnight power and how much to leave for solar the following day. Has anyone else done this for other systems?

**Reply by FrankTaylorLieder** [reply]
Hmmm.

Seems strange to have an API and give it out, but not want anyone to share details of it. Of course we can always reverse engineer it :-)

Thanks for sharing your experience.

**Comment by nathan-gs** [comment]
I put a very rough draft together in https://github.com/nathan-gs/solis-control it handles the forcechargesoc, forcedischargesoc.
